BUG: optimize: fix max function call validation for `minimize` with Powell

AtsushiSakai opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
fix a part of #12271

#### What does this implement/fix?
As reported in #12271, `minimize` with Powell method has an issue of max function call validation.
The reason for this issue is that `minimize_powell` check the max function call here once a loop:
https://github.com/scipy/scipy/blob/fbcfba22f2e4f89510f0bc1e24591a6b91204378/scipy/optimize/optimize.py#L3018-L3019
But, `_linesearch_powell` might evaluate the function multiple times.
To fix the issue, I create a new wrap function `_wrap_function_with_validation` which counts function evaluation and raises an exception when function evaluation reaches the maximum value.
We should check other methods have the same issue or not, but this PR only focuses on fixing the issue of Powell method.

rgommers commented:
This looks good, thanks @AtsushiSakai. I'd like to review/merge after the 1.7.x release split though, to not hastily introduce a last-minute issue for the release.

When looking at this again: gh-13411 is related.

AtsushiSakai commented:
I just rebased to fix conflict.

tupui commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -502,6 +502,18 @@ def test_ncg_hessp(self):
                         atol=1e-6, rtol=1e-7)
 
 
+def test_maxfev_test():
+    def cost(x):
+        return np.random.rand(1) * 1000  # never converged problem
```
Nice idea, but making it deterministic for debugging might be handy.
```suggestion
def test_maxfev_test():
    rng = np.random.default_rng(271707100830272976862395227613146332411)
    def cost(x):
        return rng.random(1) * 1000  # never converged problem
```

tupui commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -502,6 +502,18 @@ def test_ncg_hessp(self):
+        return np.random.rand(1) * 1000  # never converged problem
+
+    for imaxfev in [1, 10, 50]:
+        for method in ['Powell']:  # TODO: extend to more methods
```
What do you mean with TODO? Do you propose to do this for other methods? If yes, did you already try?

AtsushiSakai replied:
As reported in #12271, SLSQP might have a similar issue. Also, I think all optimizers should be checked by this test. 
> did you already try?

Not yet. I will try it next PR.

tupui replied:
ok let's keep this for a follow up.

tupui commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -502,6 +502,18 @@ def test_ncg_hessp(self):
+
+    for imaxfev in [1, 10, 50]:
+        for method in ['Powell']:  # TODO: extend to more methods
+            result = optimize.minimize(cost, np.random.rand(10),
```
```suggestion
            result = optimize.minimize(cost, rng.random(10),
```

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -475,6 +475,36 @@ def function_wrapper(x, *wrapper_args):
+def _wrap_scalar_function_with_validation(function, args, maxfun):
+    # wraps a minimizer function to count number of evaluations
+    # and to easily provide an args kwd.
+    ncalls = [0]
```
Why is this a list?

AtsushiSakai replied:
This is for incrementing the function call numbers in the closure function. 
The original wrapper function is using same way:
https://github.com/scipy/scipy/blob/be2027dcf11745cc28c95f088436400471d520ae/scipy/optimize/optimize.py#L454-L462

tupui replied:
Oh right to have an object and have the changes in other scopes.

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -475,6 +475,36 @@ def function_wrapper(x, *wrapper_args):
+    pass
+
+
+def _wrap_scalar_function_with_validation(function, args, maxfun):
```
This function would need to be tested since we might want to use it elsewhere. I am not certain about having it here and not in some utils.

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -475,6 +475,36 @@ def function_wrapper(x, *wrapper_args):
+
+    def function_wrapper(x, *wrapper_args):
+        if ncalls[0] >= maxfun:
+            raise _MaxFuncCallError()
```
We might want to pass a message here.

AtsushiSakai replied:
Done.

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -3015,57 +3047,60 @@ def _minimize_powell(func, x0, args=(), callback=None, bounds=None,
-            temp = fx - fx2
-            t -= delta*temp*temp
-            if t < 0.0:
+        try:
```
I am not sure about putting everything under the exception block. It might call to create another function. I am neutral on this one.

AtsushiSakai replied:
I think putting the optimization loop under the exception block is simple and has good readability because the objective function is called in many places in the optimization loop in the Powell method.

tupui reviewed (changes requested):
Thanks @AtsushiSakai. The use of a wrapper sounds like a good idea to instrument user-functions. I have a few comments.

AtsushiSakai commented:
@tupui Thank you for reviewing. I addressed your comments.
I think the CI fails are not related because I didn't touch the link error points:
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=14145&view=logs&j=fdd89725-f657-5341-7d77-3c5c44c4f5b6&t=325fc03e-e0b9-5339-adaf-099a49a14600&l=13

tupui reviewed (approved):
LGTM, thanks @AtsushiSakai. For the CI it's because you need to merge master (at least I am certain for circleCI). Could you do it quick?

AtsushiSakai commented:
I rebased it. But the same lint CI is still failed..

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -3113,57 +3146,60 @@ def _minimize_powell(func, x0, args=(), callback=None, bounds=None,
+                temp = fx - fx2
+                t -= delta*temp*temp
+                if t < 0.0:
+                    fval, x, direc1 = _linesearch_powell(func, x, direc1,
+                                                         tol=xtol * 100,
+                                                         lower_bound=lower_bound,
+                                                         upper_bound=upper_bound,
+                                                         fval=fval)
```
```suggestion
                    fval, x, direc1 = _linesearch_powell(
                        func, x, direc1,
                        tol=xtol * 100,
                        lower_bound=lower_bound,
                        upper_bound=upper_bound,
                        fval=fval
                    )
```

tupui commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -526,6 +526,36 @@ def test_ncg_hessp(self):
+    for i in range(5):
+        func(np.asarray(i))
+        assert fcalls[0] == i+1
+
+    with assert_raises(optimize.optimize._MaxFuncCallError):
```
```suggestion

    msg = "Too many function calls"
    with assert_raises(optimize.optimize._MaxFuncCallError, match=msg):
```

tupui commented on scipy/optimize/optimize.py:
```diff
@@ -475,6 +475,37 @@ def function_wrapper(x, *wrapper_args):
+            try:
+                fx = np.asarray(fx).item()
+            except (TypeError, ValueError) as e:
+                raise ValueError("The user-provided objective function "
+                                 "must return a scalar value.") from e
```
Actually this is not tested, it might be the cause of the CI failure. I am not sure since the output message is not very explicit.

tupui replied:
The CI issue is unrelated. And appears in other PRs. I will have a look at the changes I propose to merge this quick.

tupui commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -526,6 +526,37 @@ def test_ncg_hessp(self):
+        assert fcalls[0] == i+1
+
+    msg = "Too many function calls"
+    with assert_raises(optimize.optimize._MaxFuncCallError, match=msg):
+        func(np.asarray(i))  # exceeded maximum function call
+
+
```
Something like this, I will let you commit here if you agree with the suggestion.
```suggestion
    with assert_raises(optimize.optimize._MaxFuncCallError, match=msg):
        func(np.asarray(i))

    match = ("The user-provided "
             "objective function must "
             "return a scalar value.")
     with assert_raises(ValueError, match=msg):
         func(np.array([1, 1]))

```

tupui reviewed (approved):
LGTM, merging. Thanks again @AtsushiSakai
